Control the flow of time from unit test

mattias-p opened the pull request:
## Purpose

This PR introduces rpcapi.t which is intended to host unit tests for the entire RPC-API. It also adds a feature to control the flow of time from unit tests. Finally it makes use of this feature to add a unit test for the behavior of ZONEMASTER.age_reuse_previous_test.

## Context

Fixes #758 and #761.

## Changes

* The formatting commit is just there to reduce the diffs of the remaining commits.
* All questions to SQL about the current time (and some adjacent time calculations) are removed and replaced questions to Perl. It's much easier to control Perl's perception of time, and we'll be able to leverage this to be able to procedurally control time from unit tests. Fixes #758.
* Thanks to the removal of some database specific syntax we're able to deduplicate some (now identical) code in the database specific classes and move it to the common base class.
* Lastly a fix is included for #761. It takes a slightly lighter approach than what's suggested in the issue by not involving any RPC-JSON but rather hitting the RPCAPI::start_domain_test() method directly. It also deviates from the issue by letting time() stand perfectly still except for when advance_time() is called.

N.B. The database schemas still use references to NOW() in default value definitions for certain fields, but all insert queries are updated to make sure that those default values are never used. Clean up of those default value definitions is deferred to a later PR - it's more complicated because it requires an update to the database schema.

## How to test this PR

### Manual testing
Until #827 is fixed we don't have a good way to automate testing of batch jobs.
I'm opting to not expand this PR to cover #827.
Instead the changes in `DB::add_batch_job` (called from RPCAPI::add_batch_job and add-batch-job.pl) need to be tested manually.

```sh
$ zmb add_api_user --username testuser --api-key secret
$ zmb add_batch_job --username testuser --api-key secret --domain zonemaster.net
$ zmb get_batch_job_result --batch-id 1234
```

### New automatic tests
The changes in these parts are covered by the new rpcapi.t:
* DB::create_new_test
* DB::test_progress
* DB::process_unfinished_tests
* DB::process_dead_test

### Existing automatic tests
The changes in these parts are covered by test01.t:
* DB::test_progress
* DB::test_results

mattias-p commented:
I've rebased this to current develop. I plan to flesh out the unit tests some more and describe a manual test procedure for whatever remains. But it would be nice to have some eyeballs on the code.

blacksponge commented:
I have started reading through the changes, already looking good! It's nice to see new tests added for the rpcapi and deduplication of code for DBs. I will try to do some testing later on today.

mattias-p commented:
BTW, I recommend reviewing the changes commit by commit. At least up to and including "Deduplicate some code". The large commits up to that point are just changing whitespace and moving entire subroutines.

pnax commented on t/rpcapi.t:
```diff
@@ -0,0 +1,85 @@
+        advance_time( 11 );
+        my $testid2 = $rpcapi->start_domain_test( { domain => "zone1.rpcapi.example" } );
+
+        $rpcapi->{db}->test_progress( $testid1, 1 );    # mark test as started
```
Have you considered using `get_test_request`?

mattias-p replied:
I'd forgotten about that! Thanks for pointing it out, I'll update it.

mattias-p replied:
I had another look at this and realized `test_progress` has an important advantage over `get_test_request` - it lets you specify a test id. This makes the test robust against unrelated tests in the database waiting to be allocated.

pnax reviewed (commented):
I've looked over the commits and it looks very good! I'll run few tests next week.

blacksponge commented on lib/Zonemaster/Backend/DB/PostgreSQL.pm:
```diff
@@ -316,47 +329,37 @@ sub add_batch_job {
+            SET progress = 100,
+                test_end_time = ?,
+                results = ?
+            WHERE hash_id = ?",
```
What is this extra double quote for ?

mattias-p replied:
Fixed

blacksponge commented on lib/Zonemaster/Backend/DB/SQLite.pm:
```diff
@@ -266,7 +285,19 @@ sub add_batch_job {
+                hash_id,
+                domain,
+                batch_id,
+                $creation_time,
```
Shouldn't that be `creation_time` instead of `$creation_time` ?

mattias-p replied:
Fixed

blacksponge commented on lib/Zonemaster/Backend/DB/SQLite.pm:
```diff
@@ -266,7 +285,19 @@ sub add_batch_job {
+                fingerprint,
+                params,
+                undelegated
+            ) VALUES (?,?,?,?,?,?,?,?)'
```
Missing a `?` ?

mattias-p replied:
Fixed

mattias-p commented:
I tried to restart Travis for the failed test but apparently we're out of credits.

matsduf commented:
> I tried to restart Travis for the failed test but apparently we're out of credits.

I will take a look at it and request more credits. Usually we get those quickly.

matsduf commented:
> > I tried to restart Travis for the failed test but apparently we're out of credits.
> 
> I will take a look at it and request more credits. Usually we get those quickly.

Fixed.

mattias-p commented:
I'm not sure why Travis failed. The error is intermittent. I've updated the unit test to print a little more details when it does fail in this way.

mattias-p commented:
I've addressed all comments, making updates for most of them, and rebased onto the current develop branch.

However there is still one thing I'm unhappy about. The test file I've added is named rpcapi.t, but only one of three tests is really oriented around an RPCAPI procedure.

mattias-p commented:
I reframed the unit test to be about the lifecycle of tests. Currently it only covers the parts of the lifecycle that are affected by timeouts.

@blacksponge @matsduf @pnax @tgreenx Please review again.

pnax commented on lib/Zonemaster/Backend/DB.pm:
```diff
@@ -154,6 +155,28 @@ sub create_new_test {
+            SELECT hash_id
+            FROM test_results
+            WHERE fingerprint = ?
+              AND ( test_start_time IS NULL
+                 OR test_start_time >= ? )
```
Nice ! I think this should close the discussion we had in #840.

pnax reviewed (approved):
This is a very nice improvement!

mattias-p commented:
I rebased and resolved the conflict in test01.t.

pnax reviewed (commented):
The new `t/lifecycle.t` test file does not drop/create the tables. #947 changed the way we create the tables. Tables creation and removal need to be done within each test plan.

mattias-p commented:
* I updated lifecycle.t to drop tables. I also refactored test01.t because DRY.
* And I refactored the code to create the schema as well for cleanliness reasons.

pnax commented on t/lifecycle.t:
```diff
@@ -0,0 +1,116 @@
+
+use Test::More tests => 2;
+use Test::Exception;
+use Test::NoWarnings;
```
It seems that PostgreSQL emits warnings when trying to DROP a table that does not exist. This makes the tests in `t/lifecycle.t` fail.

mattias-p replied:
Good catch!

mattias-p commented:
* I updated drop_tables() to temporarily set the minimum message level.
* I also replaced a `diag` with a `note` to get prettier output from `prove`.

blacksponge reviewed (approved):
Looks good to me!

pnax commented on lib/Zonemaster/Backend/DB.pm:
```diff
@@ -316,6 +340,37 @@ sub process_unfinished_tests {
+            $self->format_time( time() - $test_run_timeout ),
+            $queue_label,
+        );
+        return $sth;
```
Have you considered having only one `return` statement in this function?

mattias-p replied:
Yeah, I really think this pattern:
```perl
if ( ... ) {
    return ...;
} else {
    return ...;
}
```
is better than this pattern:
```perl
my $z;
if ( ... ) {
    $z = ...;
} else {
    $z = ...;
}
return $z;
```

mattias-p replied:
I guess it could be useful to consider a third pattern here as well:
```perl
...
if ( ... ) {
    return ...;
}
...
return ...;
```
To me this pattern is problematic because it's easy misunderstand the structure of the control flow if you don't spot the return statement in the middle of things. But I don't think the problem is the number of `return` statements.

It's different when **all** branches of an `if / elsif / else` construct return (or throw). In this case the `if / elsif / else` construct is the last thing that happens in the procedure, so there's no code after it where you can get confused about the control flow structure.
